From 1a5144d8650cab3c60427733d9a57f87d269884a Mon Sep 17 00:00:00 2001
From: Arun Ravindran <arun.ravindran@intel.com>
Date: Tue, 21 May 2013 16:13:40 +0300
Subject: [PORT FROM MAIN] ril: Adding missing Telephony APIs for Lenovo

BZ: 95201

[DO NOT UPSTREAM]

The missing apis, causes compilation failure, when the
corresponding feature is enabled.

Change-Id: I8649e13fcaa051fedf8a3a28acc0b8a1592b27ae
Orig-Change-Id: I19184ce401f9c9b96594e3102dd757c299cb326d
Orig-Change-Id: I23d8e44f7edba69e4d5fb4735713357019b53fd9
Orig-Change-Id: Id160151a361561d40cb20932848bf4435c2eddfb
Signed-off-by: Arun Ravindran <arun.ravindran@intel.com>
---
 include/telephony/ril.h     |   84 +++++++++++++++++++++++++++++++++++++++++++
 libril/Android.mk           |   32 ++++++++++++++++
 libril/ril.cpp              |   31 ++++++++++++++++
 libril/ril_commands.h       |   16 ++++++++
 libril/ril_unsol_commands.h |    3 ++
 5 files changed, 166 insertions(+), 0 deletions(-)

diff --git a/include/telephony/ril.h b/include/telephony/ril.h
index 82e84c8..b528312 100644
--- a/include/telephony/ril.h
+++ b/include/telephony/ril.h
@@ -529,6 +529,12 @@ typedef struct
   int              pin1_replaced;   /* applicable to USIM, CSIM & ISIM */
   RIL_PinState     pin1;
   RIL_PinState     pin2;
+#if defined(M2_PIN_RETRIES_FEATURE_ENABLED)
+  int pin1_num_retries;
+  int puk1_num_retries;
+  int pin2_num_retries;
+  int puk2_num_retries;
+#endif // M2_PIN_RETRIES_FEATURE_ENABLED
 } RIL_AppStatus;
 
 /* Deprecated, use RIL_CardStatus_v6 */
@@ -3518,6 +3524,69 @@ typedef struct {
 // "response" is a const RIL_SIM_IO_Response *
 #define RIL_REQUEST_SIM_TRANSMIT_CHANNEL 114
 
+#if defined(M2_VT_FEATURE_ENABLED)
+
+/**
+ * RIL_REQUEST_HANGUP_VT
+ *
+ * Hang-up the current GSM/UMTS call of the MT, and set the cause.
+ * It can be used when the user would like to fall back an
+ * incoming VT call to VOICE call.
+ *
+ * "data" is int *
+ * ((int *)data)[0] is cause value
+ *
+ * "response" is NULL
+ *
+ * Valid errors:
+ *  SUCCESS
+ *  GENERIC_FAILURE
+ *
+ */
+#define RIL_REQUEST_HANGUP_VT 113
+
+/**
+ * RIL_REQUEST_DIAL_VT
+ *
+ * Initiate VT call.
+ *
+ * "data" is const RIL_Dial *
+ *
+ * "response" is NULL
+ *
+ * Valid errors:
+ *  SUCCESS
+ *  RADIO_NOT_AVAILABLE (radio resetting)
+ *  GENERIC_FAILURE
+ *
+ */
+#define RIL_REQUEST_DIAL_VT 114
+
+#endif  //(M2_VT_FEATURE_ENABLED)
+
+#if defined(M2_GET_SIM_SMS_STORAGE_ENABLED)
+
+/**
+ * RIL_REQUEST_GET_SIM_SMS_STORAGE
+ *
+ * Get the current memory storage information of SMS message in SIM card.
+ *
+ * "data" is NULL
+ *
+ * "response" is const int *
+ * ((const int *)response)[0] is current stored SIM SMS number
+ * ((const int *)response)[1] is total SIM SMS number
+ *
+ * Valid errors:
+ * SUCCESS, on success
+ * RADIO_NOT_AVAILABLE (radio resetting)
+ * GENERIC_FAILURE, if an error occurred
+ */
+#define RIL_REQUEST_GET_SIM_SMS_STORAGE 115
+
+#endif // M2_GET_SIM_SMS_STORAGE_ENABLED
+
+
 /***********************************************************************/
 
 
@@ -4008,6 +4077,21 @@ typedef struct {
  */
 #define RIL_UNSOL_CELL_INFO_LIST 1036
 
+#if defined(M2_CALL_FAILED_CAUSE_FEATURE_ENABLED)
+
+/**
+ * RIL_UNSOL_CALL_FAILED_CAUSE
+ *
+ * Call failed cause report.
+ *
+ * "data" is int *
+ * ((int *)data)[0] is call id
+ * ((int *)data)[1] is failed cause, refer to TS 24.008 Table 10.5.123
+ */
+#define RIL_UNSOL_CALL_FAILED_CAUSE 1037
+
+
+#endif // (M2_CALL_FAILED_CAUSE_FEATURE_ENABLED)
 /***********************************************************************/
 
 
diff --git a/libril/Android.mk b/libril/Android.mk
index b099d4b..4c7a142 100644
--- a/libril/Android.mk
+++ b/libril/Android.mk
@@ -16,6 +16,22 @@ LOCAL_SHARED_LIBRARIES := \
 
 LOCAL_CFLAGS :=
 
+ifeq ($(strip $(M2_VT_FEATURE_ENABLED)),true)
+LOCAL_CFLAGS += -DM2_VT_FEATURE_ENABLED
+endif
+
+ifeq ($(strip $(M2_CALL_FAILED_CAUSE_FEATURE_ENABLED)),true)
+LOCAL_CFLAGS += -DM2_CALL_FAILED_CAUSE_FEATURE_ENABLED
+endif
+
+ifeq ($(strip $(M2_PIN_RETRIES_FEATURE_ENABLED)),true)
+LOCAL_CFLAGS += -DM2_PIN_RETRIES_FEATURE_ENABLED
+endif
+
+ifeq ($(strip $(M2_GET_SIM_SMS_STORAGE_ENABLED)),true)
+LOCAL_CFLAGS += -DM2_GET_SIM_SMS_STORAGE_ENABLED
+endif
+
 LOCAL_MODULE:= libril
 
 LOCAL_LDLIBS += -lpthread
@@ -37,6 +53,22 @@ LOCAL_STATIC_LIBRARIES := \
 
 LOCAL_CFLAGS :=
 
+ifeq ($(strip $(M2_VT_FEATURE_ENABLED)),true)
+LOCAL_CFLAGS += -DM2_VT_FEATURE_ENABLED
+endif
+
+ifeq ($(strip $(M2_CALL_FAILED_CAUSE_FEATURE_ENABLED)),true)
+LOCAL_CFLAGS += -DM2_CALL_FAILED_CAUSE_FEATURE_ENABLED
+endif
+
+ifeq ($(strip $(M2_PIN_RETRIES_FEATURE_ENABLED)),true)
+LOCAL_CFLAGS += -DM2_PIN_RETRIES_FEATURE_ENABLED
+endif
+
+ifeq ($(strip $(M2_GET_SIM_SMS_STORAGE_ENABLED)),true)
+LOCAL_CFLAGS += -DM2_GET_SIM_SMS_STORAGE_ENABLED
+endif
+
 LOCAL_MODULE:= libril_static
 
 LOCAL_LDLIBS += -lpthread
diff --git a/libril/ril.cpp b/libril/ril.cpp
index 8113d76..3f81450 100644
--- a/libril/ril.cpp
+++ b/libril/ril.cpp
@@ -2353,6 +2353,29 @@ static void sendSimStatusAppInfo(Parcel &p, int num_apps, RIL_AppStatus appStatu
             p.writeInt32(appStatus[i].pin1_replaced);
             p.writeInt32(appStatus[i].pin1);
             p.writeInt32(appStatus[i].pin2);
+#if defined(M2_PIN_RETRIES_FEATURE_ENABLED)
+            p.writeInt32(appStatus[i].pin1_num_retries);
+            p.writeInt32(appStatus[i].puk1_num_retries);
+            p.writeInt32(appStatus[i].pin2_num_retries);
+            p.writeInt32(appStatus[i].puk2_num_retries);
+            appendPrintBuf("%s[app_type=%d,app_state=%d,perso_substate=%d,\
+                    aid_ptr=%s,app_label_ptr=%s,pin1_replaced=%d,pin1=%d,pin2=%d,\
+                    pin1_num_retries=%d,puk1_num_retries=%d,pin2_num_retries=%d,\
+                    puk2_num_retries=%d],",
+                    printBuf,
+                    appStatus[i].app_type,
+                    appStatus[i].app_state,
+                    appStatus[i].perso_substate,
+                    appStatus[i].aid_ptr,
+                    appStatus[i].app_label_ptr,
+                    appStatus[i].pin1_replaced,
+                    appStatus[i].pin1,
+                    appStatus[i].pin2,
+                    appStatus[i].pin1_num_retries,
+                    appStatus[i].puk1_num_retries,
+                    appStatus[i].pin2_num_retries,
+                    appStatus[i].puk2_num_retries);
+#else
             appendPrintBuf("%s[app_type=%d,app_state=%d,perso_substate=%d,\
                     aid_ptr=%s,app_label_ptr=%s,pin1_replaced=%d,pin1=%d,pin2=%d],",
                     printBuf,
@@ -2364,6 +2387,7 @@ static void sendSimStatusAppInfo(Parcel &p, int num_apps, RIL_AppStatus appStatu
                     appStatus[i].pin1_replaced,
                     appStatus[i].pin1,
                     appStatus[i].pin2);
+#endif // M2_PIN_RETRIES_FEATURE_ENABLED
         }
         closeResponse;
 }
@@ -3650,6 +3674,13 @@ requestToString(int request) {
         case RIL_UNSOL_RIL_CONNECTED: return "UNSOL_RIL_CONNECTED";
         case RIL_UNSOL_VOICE_RADIO_TECH_CHANGED: return "UNSOL_VOICE_RADIO_TECH_CHANGED";
         case RIL_UNSOL_CELL_INFO_LIST: return "UNSOL_CELL_INFO_LIST";
+#if defined(M2_VT_FEATURE_ENABLED)
+        case RIL_REQUEST_HANGUP_VT: return "HANGUP_VT";
+        case RIL_REQUEST_DIAL_VT: return "DIAL_VT";
+#endif
+#if defined(M2_GET_SIM_SMS_STORAGE_ENABLED)
+        case RIL_REQUEST_GET_SIM_SMS_STORAGE: return "RIL_REQUEST_GET_SIM_SMS_STORAGE";
+#endif
         default: return "<unknown request>";
     }
 }
diff --git a/libril/ril_commands.h b/libril/ril_commands.h
index 7577ed1..082f431 100644
--- a/libril/ril_commands.h
+++ b/libril/ril_commands.h
@@ -129,3 +129,19 @@
     {RIL_REQUEST_SIM_OPEN_CHANNEL, dispatchString, responseInts},
     {RIL_REQUEST_SIM_CLOSE_CHANNEL, dispatchInts, responseVoid},
     {RIL_REQUEST_SIM_TRANSMIT_CHANNEL, dispatchSIM_IO, responseSIM_IO}
+#if defined(M2_VT_FEATURE_ENABLED)
+    ,
+    {RIL_REQUEST_HANGUP_VT, dispatchInts, responseVoid},
+    {RIL_REQUEST_DIAL_VT, dispatchDial, responseVoid}
+#else
+    ,
+    {0, NULL, NULL},
+    {0, NULL, NULL}
+#endif
+#if defined(M2_GET_SIM_SMS_STORAGE_ENABLED)
+    ,
+    {RIL_REQUEST_GET_SIM_SMS_STORAGE, dispatchVoid, responseInts}
+#else
+    ,
+    {0, NULL, NULL}
+#endif
diff --git a/libril/ril_unsol_commands.h b/libril/ril_unsol_commands.h
index 0711225..2307904 100644
--- a/libril/ril_unsol_commands.h
+++ b/libril/ril_unsol_commands.h
@@ -51,3 +51,6 @@
     {RIL_UNSOL_RIL_CONNECTED, responseInts, WAKE_PARTIAL},
     {RIL_UNSOL_VOICE_RADIO_TECH_CHANGED, responseInts, WAKE_PARTIAL},
     {RIL_UNSOL_CELL_INFO_LIST, responseCellInfoList, WAKE_PARTIAL},
+#if defined(M2_CALL_FAILED_CAUSE_FEATURE_ENABLED)
+    {RIL_UNSOL_CALL_FAILED_CAUSE, responseInts, WAKE_PARTIAL},
+#endif
-- 
1.7.4.1

